contrib/babelfishpg_tsql/src/pltsql_coerce.c

```diff
@@ -1070,15 +1070,55 @@ validate_special_function(char *func_nsname, char *func_name, int nargs, Oid *in
 
 	/* If function is not a special function no additional handling required */
 	if (special_func == NULL)
+	{
+		/* report error for case when NULL casted to different datatypes and passed as 2nd or 3rd argument of SUBSTRING() function */
+		if (strlen(func_name) == 9 && strncmp(func_name, "substring", 9) == 0)
+		{
+			for (int i = 1; i < nargs; i++)
+			{
+				if (input_typeids[i] != UNKNOWNOID)
+				{
+					Node *arg = (Node *) lfirst(list_nth_cell(fargs, i));
+					/* Throw error when input is constant and NULL */
+					if (IsA(arg, Const) && ((Const *)arg)->constisnull)
+					{
+						if (common_utility_plugin_ptr == NULL)
+							ereport(ERROR,
+									(errcode(ERRCODE_INTERNAL_ERROR),
+										errmsg("Failed to find common utility plugin.")));
+
+						if ((*common_utility_plugin_ptr->get_tsql_datatype_oid)("int") != input_typeids[i]
+							&& (*common_utility_plugin_ptr->get_tsql_datatype_oid)("bigint") != input_typeids[i]
+							&& (*common_utility_plugin_ptr->get_tsql_datatype_oid)("smallint") != input_typeids[i]
+							&& (*common_utility_plugin_ptr->get_tsql_datatype_oid)("tinyint") != input_typeids[i])
```

Comment:
Suggestion: Finding tsql typname by oid and then string comparsion will be more efficient here.

Reply:
nack, here get_tsql_datatype_oid works on string comparison so it not "faster" instead what i suggested will be faster.

> less prone to vulnerability than string comparison

Can you explain how that can lead to vulnerability?

Reply:
Got it, Modified